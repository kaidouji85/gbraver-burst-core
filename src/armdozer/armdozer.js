// @flow

/** アームドーザID */
export type ArmDozerId = string;

// TODO バーストの種類を追加する

/** アームドーザ基本情報 */
export type Armdozer = {
  /** アームドーザID */
  id: ArmDozerId,
  /** 名前 */
  name: string,
  /** 最大HP */
  maxHp: number,
  /** 最大バッテリー */
  maxBattery: number,
  /** 攻撃 */
  power: number,
  /** スピード */
  speed: number,
  /** アームドーザの外見を指定する */
  appearance: string,
};
