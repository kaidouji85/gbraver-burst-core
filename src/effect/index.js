// @flow

import type {InputCommand} from "./input-command/input-command";
import type {Battle} from "./battle/battle";
import type {TurnChange} from "./turn-change/turn-change";
import type {StartGame} from "./start-game/start-game";
import type {BurstEffect} from "./burst/burst-effect";
import type {GameEnd} from "./game-end/game-end";
import type {BatteryDeclaration} from "./battery-declaration/battery-declaration";
import type {Reflect} from "./reflect/reflect";
import type {UpdateRemainingTurn} from "./update-remaning-turn/update-remaining-turn";

/** ゲーム中に発生する効果をまとめたもの */
export type Effect =
  StartGame |
  GameEnd |
  InputCommand |
  Reflect |
  BatteryDeclaration |
  Battle |
  TurnChange |
  BurstEffect |
  UpdateRemainingTurn;
