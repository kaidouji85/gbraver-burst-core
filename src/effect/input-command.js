// @flow

import * as R from 'ramda';
import type {PlayerId} from "../player/player";
import type {Command} from "../command/command";
import type {ArmdozerState} from "../game-state/armdozer-state";

/** コマンド入力 */
export type InputCommand = {
  name: 'InputCommand',
  players: EnableCommand[]
};

/** プレイヤー毎の入力可能コマンドをまとめたもの */
export type EnableCommand = {
  playerId: PlayerId,
  command: Command[]
};

/**
 * プレイヤーの状態から入力可能なコマンドを計算する
 *
 * @param player プレイヤーの状態
 * @return 入力可能なコマンド
 */
export function getEnableCommandList(armdozer: ArmdozerState): Command[] {
  return [
    ...getBatteryCommand(armdozer),
    ...getBurstCommand(armdozer)
  ];
}

/**
 * アームドーザの状態から、バッテリー系コマンドで使えるものを返す
 * 使えるコマンドがない場合は空配列を返す
 *
 * @param armdozer アームドーザの状態
 * @return 計算結果
 */
function getBatteryCommand(armdozer: ArmdozerState): Command[] {
  return R.range(0, armdozer.maxBattery + 1)
    .filter(v => v <= armdozer.battery)
    .map(v => ({type: 'BATTERY_COMMAND', battery: v}));
}

/**
 * アームドーザの状態から、バースト系コマンドで使えるものを返す
 * 使えるコマンドがない場合は空配列を返す
 *
 * @param armdozer アームドーザの状態
 * @return 計算結果
 */
function getBurstCommand(armdozer: ArmdozerState): Command[] {
  return armdozer.enableBurst ? [{type: 'BURST_COMMAND'}] : [];
}