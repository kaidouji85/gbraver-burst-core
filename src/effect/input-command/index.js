// @flow

import type {GameState} from "../../state/game-state";
import type {NoChoice, Selectable} from "./input-command";
import type {Command, PlayerCommand} from "../../command/command";
import type {PlayerState} from "../../state/player-state";
import {isSelectableCommand} from "./is-selectable-command";
import {selectableBatteryCommand} from "./selectable-battery-command";
import {selectableBurstCommand} from "./selectable-burst-command";

/**
 * コマンド入力フェイズのステートを生成する
 * ゲーム開始時、誰もバーストコマンドを選択したなかった場合、本関数でコマンド入力フェイズの効果を解決すること
 *
 * @param lastState 更新前の状態
 * @return コマンド入力状態
 */
export function inputCommand(lastState: GameState): GameState {
  return {
    ...lastState,
    effect: {
      name: 'InputCommand',
      players: lastState.players.map(v => ({
        playerId: v.playerId,
        selectable: true,
        command: [
          ...selectableBatteryCommand(v.armdozer),
          ...selectableBurstCommand(v.armdozer)
        ]
      }))
    }
  };
}

/**
 * バーストフェイズ後のコマンド入力フェイズのステートを生成する
 * 本関数はいずれかのプレイヤーがバーストコマンドを選択した場合に、実行される
 *
 * @param lastState 更新前の状態
 * @param commands バーストフェイズ発生時に各プレイヤーが選択したコマンド
 * @returns コマンド入力状態
 */
export function inputCommandAfterBurst(lastState: GameState, commands: PlayerCommand[]): GameState {
  return {
    ...lastState,
    effect: {
      name: 'InputCommand',
      players: lastState.players.map(player => {
        const playerCommand = commands.find(v => v.playerId === player.playerId);
        if (!playerCommand) {
          return selectable(player);
        }

        return isSelectableCommand(playerCommand.command)
          ? selectable(player)
          : noChoice(player, playerCommand.command);
      })
    }
  };
}

/** コマンド選択可能なケース */
function selectable(player: PlayerState): Selectable {
  return {
    playerId: player.playerId,
    selectable: true,
    command: [
      ...selectableBatteryCommand(player.armdozer)
    ]
  };
}

/** コマンド選択が不可能なケース */
function noChoice(player: PlayerState, command: Command): NoChoice {
  return {
    playerId: player.playerId,
    selectable: false,
    nextTurnCommand: command
  };
}