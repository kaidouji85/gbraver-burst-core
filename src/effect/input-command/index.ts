import type { Command } from "../../command/command";
import type { PlayerId } from "../../player/player";
import type { GameState, GameStateX } from "../../state/game-state";
import type { PlayerState } from "../../state/player-state";
import type { InputCommand, NoChoice, Selectable } from "./input-command";
import { selectableBatteryCommand } from "./selectable-battery-command";
import { selectableBurstCommand } from "./selectable-burst-command";
import { selectablePilotSkillCommand } from "./selectable-pilot-skill-command";
import {castQuickCommand} from "../../command/quick-command";

/**
 * ゲームスタート時だけに利用するInputCommand
 * InputCommandはそのターンに入力したコマンドを参照する想定だが、
 * ゲーム開始時にコマンド入力できないので、本関数を用意した
 *
 * @param lastState 最新状態
 * @return 更新結果
 */
export function gameStartInputCommand(lastState: GameState): GameState {
  return {
    ...lastState,
    effect: {
      name: "InputCommand",
      players: lastState.players.map((v) => selectable(v)),
    },
  };
}
// TODO 引数を[PlayerCommand, PlayerCommand]に変更する

/**
 * コマンド入力フェイズのステートを生成する
 *
 * @param lastState 更新前の状態
 * @param attackerId 攻撃側プレイヤーID
 * @param attackerCommand 攻撃側コマンド
 * @param defenderId 防御側プレイヤーID
 * @param defenderCommand 防御側コマンド
 * @return 更新結果
 */
export function inputCommand(
  lastState: GameState,
  attackerId: PlayerId,
  attackerCommand: Command,
  defenderId: PlayerId,
  defenderCommand: Command
): GameStateX<InputCommand> {
  const attacker = lastState.players.find((v) => v.playerId === attackerId);
  const defender = lastState.players.find((v) => v.playerId === defenderId);

  if (!attacker || !defender) {
    throw new Error("not found attacker or defender command");
  }

  const nextAttackerCommand = isNoChoice(attackerCommand, defenderCommand)
    ? noChoice(attacker, attackerCommand)
    : selectable(attacker);
  const nextDefenderCommand = isNoChoice(defenderCommand, attackerCommand)
    ? noChoice(defender, defenderCommand)
    : selectable(defender);
  const playerCommands = [nextAttackerCommand, nextDefenderCommand];
  const effect: InputCommand = {
    name: "InputCommand",
    players: playerCommands,
  };
  return { ...lastState, effect: effect };
}

/**
 * コマンド選択不可能か否かを判定する
 *
 * @param myCommand 自分のコマンド
 * @param otherCommand 相手のコマンド
 * @return {boolean}
 */
export function isNoChoice(myCommand: Command, otherCommand: Command): boolean {
  return (
    myCommand.type === "BATTERY_COMMAND" && !!castQuickCommand(otherCommand)
  );
}

/**
 * コマンド選択可能なケース
 *
 * @param player プレイヤー情報
 * @return プレイヤーが次のターンに入力可能なコマンド
 */
function selectable(player: PlayerState): Selectable {
  return {
    playerId: player.playerId,
    selectable: true,
    command: [
      ...selectableBatteryCommand(player.armdozer),
      ...selectableBurstCommand(player.armdozer),
      ...selectablePilotSkillCommand(player.pilot),
    ],
  };
}

/**
 * コマンド選択が不可能なケース
 * このターンに選択したコマンドをそのまま次も選択する
 *
 * @param player プレイヤー情報
 * @param command このターンの入力したコマンド
 * @return プレイヤーが次のターンに入力可能なコマンド
 */
function noChoice(player: PlayerState, command: Command): NoChoice {
  return {
    playerId: player.playerId,
    selectable: false,
    nextTurnCommand: command,
  };
}
