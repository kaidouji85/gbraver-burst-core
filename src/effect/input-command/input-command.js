// @flow

import type { Command } from "../../command/command";
import type { PlayerId } from "../../player/player";

/** コマンド入力 */
export type InputCommand = {
  name: "InputCommand",

  /** 各プレイヤーのコマンド */
  players: Array<Selectable | NoChoice>,
};

/**
 * コマンドが自由に選択できるケース
 * selectable = trueなら、このデータ型になる
 */
export type Selectable = {
  /** プレイヤーID */
  playerId: PlayerId,
  /** コマンドが選択できるか否かのフラグ*/
  selectable: true,
  /** プレイヤーが選択できるコマンド */
  command: Command[],
};

/**
 * コマンドが強制されているケース
 * selectable = falseなら、このデータ型になる
 */
export type NoChoice = {
  /** プレイヤーID */
  playerId: PlayerId,
  /** コマンドが選択できるか否かのフラグ*/
  selectable: false,
  /** プレイヤーが次のターンに実行するコマンド */
  nextTurnCommand: Command,
};
