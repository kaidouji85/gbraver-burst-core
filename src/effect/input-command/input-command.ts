import type { Command } from "../../command/command";
import type { PlayerId } from "../../player/player";

/** コマンドが自由に選択できるケース */
export type Selectable = Readonly<{
  /** プレイヤーID */
  playerId: PlayerId;
  /** コマンドが選択できるか否かのフラグ、自由に選択できるのでtrue固定 */
  selectable: true;
  /** プレイヤーが選択できるコマンド */
  command: Command[];
}>;

/** コマンドが強制されているケース */
export type NoChoice = Readonly<{
  /** プレイヤーID */
  playerId: PlayerId;
  /** コマンドが選択できるか否かのフラグ、コマンド強制されているのでfalse固定 */
  selectable: false;
  /** プレイヤーが次のターンに実行するコマンド */
  nextTurnCommand: Command;
}>;

/** コマンド入力 */
export type InputCommand = Readonly<{
  name: "InputCommand";
  /** 各プレイヤーのコマンド */
  players: Array<Selectable | NoChoice>;
}>;