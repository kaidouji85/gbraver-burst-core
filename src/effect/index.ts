import { z } from "zod";

import { BatteryDeclaration, BatteryDeclarationSchema } from "./battery-declaration/battery-declaration";
import { Battle, BattleSchema } from "./battle/battle";
import { BurstEffect, BurstEffectSchema } from "./burst/burst-effect";
import { GameEnd, GameEndSchema } from "./game-end/game-end";
import { InputCommand, InputCommandSchema } from "./input-command/input-command";
import { PilotSkillEffect, PilotSkillEffectSchema } from "./pilot-skill/pilot-skill-effect";
import { Reflect, ReflectSchema } from "./reflect/reflect";
import { RightItself, RightItselfSchema } from "./right-itself/right-itself";
import { StartGame, StartGameSchema } from "./start-game/start-game";
import { TurnChange, TurnChangeSchema } from "./turn-change/turn-change";
import { UpdateRemainingTurn, UpdateRemainingTurnSchema } from "./update-remaning-turn/update-remaining-turn";

/** ゲーム中に発生する効果をまとめたもの */
export type Effect =
  | StartGame
  | GameEnd
  | InputCommand
  | Reflect
  | BatteryDeclaration
  | Battle
  | TurnChange
  | BurstEffect
  | UpdateRemainingTurn
  | RightItself
  | PilotSkillEffect;

/** Effect zodスキーマ */
export const EffectSchema = z.union([
  StartGameSchema,
  GameEndSchema,
  InputCommandSchema,
  ReflectSchema,
  BatteryDeclarationSchema,
  BattleSchema,
  TurnChangeSchema,
  BurstEffectSchema,
  UpdateRemainingTurnSchema,
  RightItselfSchema,
  PilotSkillEffectSchema,
]);