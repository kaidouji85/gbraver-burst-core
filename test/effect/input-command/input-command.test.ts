import { BatteryCommand, PlayerId } from "../../../src";
import { inputCommand } from "../../../src/effect/input-command";
import { EMPTY_ARMDOZER_STATE } from "../../../src/empty/armdozer";
import { EMPTY_GAME_STATE } from "../../../src/empty/game-state";
import { EMPTY_PLAYER_STATE } from "../../../src/empty/player";

/**
 * プレイヤーを生成する
 * @param playerId プレイヤーID
 * @param battery バッテリー
 * @param enableBurst バースト発動可能か否か、trueで発動可能
 * @returns 生成されたプレイヤー
 */
const createPlayer = (
  playerId: PlayerId,
  battery: number,
  enableBurst: boolean,
) => ({
  ...EMPTY_PLAYER_STATE,
  playerId,
  armdozer: {
    ...EMPTY_ARMDOZER_STATE,
    battery,
    maxBattery: 5,
    enableBurst,
  },
});

/**
 * バッテリーコマンドを生成する
 * @param battery バッテリー値
 * @returns 生成されたバッテリーコマンド
 */
const createBatteryCommand = (battery: number): BatteryCommand => ({
  type: "BATTERY_COMMAND",
  battery,
});

test("プレイヤーの状況に応じて、選択可能なコマンドがセットされる", () => {
  const player01 = createPlayer("player01", 5, true);
  const player02 = createPlayer("player02", 3, false);
  const lastState = {
    ...EMPTY_GAME_STATE,
    activePlayerId: player01.playerId,
    players: [player01, player02],
  };

  expect(inputCommand({ lastState, noChoices: [] })).toMatchSnapshot();
});

test("効果適用フロー後のコマンド入力フェイズ効果が正しく処理される", () => {
  const player01 = createPlayer("player01", 2, true);
  const player02 = createPlayer("player02", 3, false);
  const player02Command = {
    playerId: player02.playerId,
    command: createBatteryCommand(2),
  };
  const lastState = {
    ...EMPTY_GAME_STATE,
    activePlayerId: player02.playerId,
    players: [player01, player02],
  };

  expect(
    inputCommand({
      lastState,
      noChoices: [player02Command],
    }),
  ).toMatchSnapshot("after-effect-activation");
});
