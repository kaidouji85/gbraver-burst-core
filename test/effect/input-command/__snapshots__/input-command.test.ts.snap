// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`プレイヤーの状況に応じて、選択可能なコマンドがセットされる 1`] = `
{
  "activePlayerId": "player01",
  "effect": {
    "name": "InputCommand",
    "players": [
      {
        "command": [
          {
            "battery": 0,
            "type": "BATTERY_COMMAND",
          },
          {
            "battery": 1,
            "type": "BATTERY_COMMAND",
          },
          {
            "battery": 2,
            "type": "BATTERY_COMMAND",
          },
          {
            "battery": 3,
            "type": "BATTERY_COMMAND",
          },
          {
            "battery": 4,
            "type": "BATTERY_COMMAND",
          },
          {
            "battery": 5,
            "type": "BATTERY_COMMAND",
          },
          {
            "type": "BURST_COMMAND",
          },
          {
            "type": "PILOT_SKILL_COMMAND",
          },
        ],
        "playerId": "player01",
        "selectable": true,
      },
      {
        "command": [
          {
            "battery": 0,
            "type": "BATTERY_COMMAND",
          },
          {
            "battery": 1,
            "type": "BATTERY_COMMAND",
          },
          {
            "battery": 2,
            "type": "BATTERY_COMMAND",
          },
          {
            "battery": 3,
            "type": "BATTERY_COMMAND",
          },
          {
            "type": "PILOT_SKILL_COMMAND",
          },
        ],
        "playerId": "player02",
        "selectable": true,
      },
    ],
  },
  "players": [
    {
      "armdozer": {
        "battery": 5,
        "burst": {
          "recoverBattery": 5,
          "turnStartBatteryCorrect": 1,
          "type": "RecoverBattery",
        },
        "effects": [],
        "enableBurst": true,
        "hp": 3000,
        "id": "id",
        "maxBattery": 5,
        "maxHp": 3000,
        "name": "name",
        "power": 2000,
        "speed": 2000,
      },
      "pilot": {
        "enableSkill": true,
        "id": "EMPTY_PLAYER",
        "name": "名無し",
        "skill": {
          "recoverBattery": 2,
          "type": "RecoverBatterySkill",
        },
      },
      "playerId": "player01",
    },
    {
      "armdozer": {
        "battery": 3,
        "burst": {
          "recoverBattery": 5,
          "turnStartBatteryCorrect": 1,
          "type": "RecoverBattery",
        },
        "effects": [],
        "enableBurst": false,
        "hp": 3000,
        "id": "id",
        "maxBattery": 5,
        "maxHp": 3000,
        "name": "name",
        "power": 2000,
        "speed": 2000,
      },
      "pilot": {
        "enableSkill": true,
        "id": "EMPTY_PLAYER",
        "name": "名無し",
        "skill": {
          "recoverBattery": 2,
          "type": "RecoverBatterySkill",
        },
      },
      "playerId": "player02",
    },
  ],
}
`;

exports[`効果適用フロー後のコマンド入力フェイズ効果が正しく処理される: after-effect-activation 1`] = `
{
  "activePlayerId": "player02",
  "effect": {
    "name": "InputCommand",
    "players": [
      {
        "nextTurnCommand": {
          "battery": 2,
          "type": "BATTERY_COMMAND",
        },
        "playerId": "player02",
        "selectable": false,
      },
      {
        "command": [
          {
            "battery": 0,
            "type": "BATTERY_COMMAND",
          },
          {
            "battery": 1,
            "type": "BATTERY_COMMAND",
          },
          {
            "battery": 2,
            "type": "BATTERY_COMMAND",
          },
          {
            "type": "BURST_COMMAND",
          },
          {
            "type": "PILOT_SKILL_COMMAND",
          },
        ],
        "playerId": "player01",
        "selectable": true,
      },
    ],
  },
  "players": [
    {
      "armdozer": {
        "battery": 2,
        "burst": {
          "recoverBattery": 5,
          "turnStartBatteryCorrect": 1,
          "type": "RecoverBattery",
        },
        "effects": [],
        "enableBurst": true,
        "hp": 3000,
        "id": "id",
        "maxBattery": 5,
        "maxHp": 3000,
        "name": "name",
        "power": 2000,
        "speed": 2000,
      },
      "pilot": {
        "enableSkill": true,
        "id": "EMPTY_PLAYER",
        "name": "名無し",
        "skill": {
          "recoverBattery": 2,
          "type": "RecoverBatterySkill",
        },
      },
      "playerId": "player01",
    },
    {
      "armdozer": {
        "battery": 3,
        "burst": {
          "recoverBattery": 5,
          "turnStartBatteryCorrect": 1,
          "type": "RecoverBattery",
        },
        "effects": [],
        "enableBurst": false,
        "hp": 3000,
        "id": "id",
        "maxBattery": 5,
        "maxHp": 3000,
        "name": "name",
        "power": 2000,
        "speed": 2000,
      },
      "pilot": {
        "enableSkill": true,
        "id": "EMPTY_PLAYER",
        "name": "名無し",
        "skill": {
          "recoverBattery": 2,
          "type": "RecoverBatterySkill",
        },
      },
      "playerId": "player02",
    },
  ],
}
`;
